core/state/statedb.go

```diff
@@ -728,6 +728,33 @@ func (s *StateDB) Copy() *StateDB {
 	if s.prefetcher != nil {
 		state.prefetcher = s.prefetcher.copy()
 	}
+	if s.snaps != nil {
+		// In order for the miner to be able to use and make additions
+		// to the snapshot tree, we need to copy that aswell.
+		// Otherwise, any block mined by ourselves will cause gaps in the tree,
+		// and force the miner to operate trie-backed only
+		state.snaps = s.snaps
+		state.snap = s.snap
+		// deep copy needed
+		state.snapDestructs = make(map[common.Hash]struct{})
+		for k, v := range s.snapDestructs {
+			state.snapDestructs[k] = v
+		}
+		state.snapAccounts = make(map[common.Hash][]byte)
+		for k, v := range s.snapAccounts {
+			state.snapAccounts[k] = v
```

Comment:
Q: do we need to also copy the `v` byteslice?

Reply:
I don't think so. Important to know though that these values get inserted verbatim into a diff layer and whe nretrieving storage at least those get returned again verbatim. So `snapshot.Storage(0x)[2] = 2` would modify it. That said, I don't know of any reason why you'd do such a thing :D

The snapshot explicitly warns:

```
// Note the returned slot is not a copy, please don't modify it.
func (dl *diffLayer) Storage(accountHash, storageHash common.Hash) ([]byte, error) {
```